### PR TITLE
Optimized deepseek-v3/r1 model performance on mxfp4 run

### PR DESCRIPTION
**Pull request by kkHuang-amd**
## Motivation

In order to decrease the activation tensor quantized overhead, fused the quantized behavior into the different ops (activation, layernorm, gemm, flatten)

## Benchmarking and Profiling
Test below commands, we got about 10% throughput increased.
**Server:**
SGLANG_USE_AITER=1 python3 -m sglang.launch_server --model-path ams/DeepSeek-R1-WMXFP4-Preview --tp 8 --trust-remote-code --chunked-prefill-size 131072

**Client**
python3 -m sglang.bench_serving --dataset-name random --random-range-ratio 1 --num-prompt 500 --random-input 3200 --random-output 800 --max-concurrency 128

[Before this commit]
image

[Apply this commit]
image

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

File: python/sglang/srt/layers/communicator.py [hunk 1]
```diff
@@ -43,13 +43,19 @@
...
 _is_flashinfer_available = is_flashinfer_available()
 _is_sm100_supported = is_cuda() and is_sm100_supported()
+_use_aiter = get_bool_env_var("SGLANG_USE_AITER") and is_hip() 
+
+if _use_aiter:
+    from aiter.ops.triton.fused_mxfp4_quant import fused_rms_mxfp4_quant
 
 FUSE_ALLREDUCE_MAX_BATCH_SIZE = 2048
 
+from aiter.ops.triton.fused_mxfp4_quant import fused_rms_mxfp4_quant
```
Comment (HaiShaw) [inline python/sglang/srt/layers/communicator.py]: removable

File: python/sglang/srt/layers/quantization/quark/schemes/quark_w4a4_mxfp4.py [hunk 2]
```diff
@@ -20,6 +21,63 @@
 
 OCP_MX_BLOCK_SIZE = 32
 
+def b_mxfp4_to_f32(x):
```
Comment (HaiShaw) [inline python/sglang/srt/layers/quantization/quark/schemes/quark_w4a4_mxfp4.py]: consider to add comment to the 2 functions below, seems just one different.
consider to move `mxfp4_list` definition out of local.

File: python/sglang/srt/layers/quantization/quark/schemes/quark_w4a4_mxfp4.py [hunk 3]
```diff
@@ -108,11 +165,53 @@ def apply_weights(
...
 
-        out = gemm_afp4wfp4(x_q, layer.weight, x_s, layer.weight_scale, out_dtype, y)
+        #x_q, x_s = dynamic_mxfp4_quant(x)
+        #y = torch.empty(
+        #    x_q.shape[0], layer.weight.shape[0], device=x_q.device, dtype=out_dtype
+        #)
+
+        #out = gemm_afp4wfp4(x_q, layer.weight, x_s, layer.weight_scale, out_dtype, y)
+
+        #return out
```
[the remark this answers is not in the document]
Comment (HaiShaw) [reply python/sglang/srt/layers/quantization/quark/schemes/quark_w4a4_mxfp4.py]: remove

File: python/sglang/srt/layers/activation.py [hunk 4]
```diff
@@ -49,6 +49,7 @@
     from sgl_kernel import gelu_and_mul, gelu_tanh_and_mul, silu_and_mul
 elif _is_hip:
     from sgl_kernel import gelu_and_mul, gelu_quick, gelu_tanh_and_mul, silu_and_mul
+    from aiter.ops.triton.activation import act_mul_and_mxfp4_quant
```
Comment (HaiShaw) [inline python/sglang/srt/layers/activation.py]: It supports what kinds of activations?

File: python/sglang/srt/layers/activation.py [hunk 5]
```diff
@@ -57,18 +58,21 @@
 
 
 class SiluAndMul(CustomOp):
-    def forward_native(self, x: torch.Tensor) -> torch.Tensor:
+    def forward_native(self, x: torch.Tensor, fused_mxfp4_prequant: Optional[bool] = False) -> torch.Tensor:
```
Comment (HaiShaw) [inline python/sglang/srt/layers/activation.py]: `fused_mxfp4_prequant` -> `fused_mxfp4_quant`

File: python/sglang/srt/layers/activation.py [hunk 6]
```diff
@@ -57,18 +58,21 @@
...
         return F.silu(x[..., :d]) * x[..., d:]
 
-    def forward_cuda(self, x: torch.Tensor) -> torch.Tensor:
-        d = x.shape[-1] // 2
-        output_shape = x.shape[:-1] + (d,)
-        out = torch.empty(output_shape, dtype=x.dtype, device=x.device)
-        silu_and_mul(x, out)
+    def forward_cuda(self, x: torch.Tensor, fused_mxfp4_prequant: Optional[bool] = False) -> torch.Tensor:
+        if fused_mxfp4_prequant:
+            out = act_mul_and_mxfp4_quant(x, "silu")
```
Comment (HaiShaw) [inline python/sglang/srt/layers/activation.py]: Use adjusted flag name: `fused_mxfp4_quant`.
We should define interface to be more generic, vs. a flag, as function interface isn't designed to be mxfp4 specific.
We can check add an Optional dtype of output, base on its type to activate the extra logic necessary.
And - where is `act_mul_and_mxfp4_quant(x, "gelu")` in this PR?

**Review by HaiShaw — CHANGES_REQUESTED** [review]

File: python/sglang/srt/layers/activation.py (hunk 4 above)
Comment (kkHuang-amd) [reply python/sglang/srt/layers/activation.py]: gelu and silu supported

File: python/sglang/srt/models/deepseek_v2.py [hunk 7]
```diff
@@ -2344,6 +2447,32 @@ def post_load_weights(self, is_nextn=False, weight_names=None):
...
+
+            if _use_aiter and self.quant_config.get_name() == "quark":
+                #dynamic quant for mxfp4
+                if w.dtype == torch.bfloat16:
+                    w_kc, w_s_kc = b_dynamic_mxfp4_quant(w_kc.transpose(-2,-1))
+                    w_kc = w_kc.transpose(-2,-1)
+                    w_s_kc = w_s_kc.transpose(-2,-1)
+                    w_vc, w_s_vc = b_dynamic_mxfp4_quant(w_vc)
+                    self_attn.w_scale_k = w_s_kc.transpose(1, 2).contiguous().transpose(1, 2)
+                    self_attn.w_scale_v = w_s_vc.contiguous().transpose(1, 2)
```
Comment (BowenBao) [inline python/sglang/srt/models/deepseek_v2.py]: Thank you for adding this support. Depending on accuracy requirements potentially a flag to turn this on/off would be helpful.

Do you have the before and after gsm8k scores?

File: python/sglang/srt/layers/quantization/quark/utils.py [hunk 8]
```diff
@@ -105,3 +109,75 @@ def _is_equal_or_regex_match(
...
+    mxfp4_in_f32 = torch.tensor(mxfp4_list, dtype=torch.float32, device="cuda")
+    return mxfp4_in_f32[x.long()]
+
+
+def e8m0_to_f32(x):
+    x_f32 = 2 ** ((x.to(torch.float32)) - 127)
+    x_f32[x_f32 == 128] = float("nan")
+    return x_f32
+
+
```
Comment (HaiShaw) [inline python/sglang/srt/layers/quantization/quark/utils.py:166]: Please add comments/explanation to code block below

File: python/sglang/srt/layers/quantization/quark/schemes/quark_w4a4_mxfp4.py [hunk 9]
```diff
@@ -93,26 +85,51 @@ def apply_weights(
...
-        y = torch.empty(
-            x_q.shape[0], layer.weight.shape[0], device=x_q.device, dtype=out_dtype
+        # This path does not have support for bias currently
+        assert bias is None, "bias is not supported"
+
+        three_d = False
+        x_s = None
+        y = None
+        if isinstance(x, tuple):
+            assert len(x) in [
```
Comment (HaiShaw) [inline python/sglang/srt/layers/quantization/quark/schemes/quark_w4a4_mxfp4.py:95]: whitespace, tab issues

**Review by HaiShaw — COMMENTED** [review]
some to change

File: python/sglang/srt/layers/quantization/quark/utils.py [hunk 10]
```diff
@@ -5,6 +5,10 @@
 from types import MappingProxyType
 from typing import Any, Optional
 
+import torch
+from aiter.ops.triton.quant import dynamic_mxfp4_quant
```
Comment (HaiShaw) [inline python/sglang/srt/layers/quantization/quark/utils.py:9]: need to be protected with `_use_aiter`?

File: python/sglang/srt/layers/quantization/quark/utils.py [hunk 11]
```diff
@@ -105,3 +109,75 @@ def _is_equal_or_regex_match(
...
+    return mxfp4_in_f32[x.long()]
+
+
+def e8m0_to_f32(x):
+    x_f32 = 2 ** ((x.to(torch.float32)) - 127)
+    x_f32[x_f32 == 128] = float("nan")
+    return x_f32
+
+
+def quark_post_load_weights(self_attn: nn.Module, w: torch.Tensor, quant_format: str):
```
Comment (HaiShaw) [inline python/sglang/srt/layers/quantization/quark/utils.py:167]: Let's add a few comments below for the flow

**Review by HaiShaw — APPROVED** [review]
LGTM

**Comment by HaiShaw** [comment]
cc @saienduri
